optaplanner-examples/src/test/java/org/optaplanner/examples/nqueens/solver/score/NQueensConstraintProviderTest.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.nqueens.solver.score;
+
+import org.junit.Test;
+import org.optaplanner.examples.nqueens.domain.Column;
+import org.optaplanner.examples.nqueens.domain.NQueens;
+import org.optaplanner.examples.nqueens.domain.Queen;
+import org.optaplanner.examples.nqueens.domain.Row;
+import org.optaplanner.test.impl.score.stream.ConstraintVerifier;
+import org.optaplanner.test.impl.score.stream.SingleConstraintVerifier;
+
+public class NQueensConstraintProviderTest {
+
+    private final NQueensConstraintProvider constraintProvider = new NQueensConstraintProvider();
+    private final ConstraintVerifier<NQueens> constraintVerifier =
+            ConstraintVerifier.createFor(NQueens.class, Queen.class);
+
+    @Test
+    public void testSingleConstraint() {
+        Column column1 = new Column();
+        column1.setId(0L);
+        column1.setIndex(0);
+        Row row = new Row();
+        row.setId(0L);
+        row.setIndex(0);
+        Queen queen1 = new Queen();
+        queen1.setId(0L);
+        queen1.setRow(row);
+        queen1.setColumn(column1);
+        // One queen
+        SingleConstraintVerifier<NQueens> horizontalConflictConstraintVerifier =
+                constraintVerifier.forConstraint(constraintProvider::horizontalConflict);
+        horizontalConflictConstraintVerifier.givenFacts(queen1, row, column1)
+                .expectImpact("No horizontal conflicts with just one queen.", 0);
```

Comment:
What is this number? Is it the sum of `matchWeight`s from all matches of this constraint?

Reply:
I'm thinking about two use cases:
1. I only want to verify how many times the constraint matched and I don't care about the match weights.
2. I want to verify the match weight was calculated correctly.

I think that use case 2 is more important and it also indirectly verifies the match count. So the question is: is use case 1 valid? Should it be supported as well? My proposed answer is yes. If it is, then the API should discern between these two verification types.

Reply:
Feel free to disagree. I probably shouldn't have looked at the implementation where the impact number is calculated as a sum of matches. The weight is disregarded. That made me ask the original question in this thread.

Now I think verifying the total match weight is enough.

Reply:
That's a good question.

I believe that the constraint is more than just a number of matches. The match weight is calculated, and in some cases it's calculated with a relatively complicated code. Therefore it needs to be tested.

You know I was originally arguing that we should evaluate the resulting score, not just the weights. I have since changed my mind on that. But I think that further reducing the problem to just match counts would be over-simplifying.

Reply:
I'm thinking the same.

I still think verifying the match count *could be* useful in some cases but we can think about that later.